### PR TITLE
Account for tape rule status in MSRulecleaner.

### LINKED ISSUE #10027: Account for Tape Rule Status in MSRuleCleaner
**Impact of the new feature**
Currently MSRuleCleaner removes rules without consideration of the status of tape rules made in MSOutput.  Here I request MSRuleCleaner pay attention to tape rule status to avoid deleting the source of unapproved tape rules.

**Is your feature request related to a problem? Please describe.**
Yes!  If MSRuleCleaner deletes a disk rule that is the only protection for source data for a tape rule that hasn't been approved yet, then Rucio may delete that data before the rule gets approved.

**Describe the solution you'd like**
In the Rucio Transition meeting, we discussed the possibilty of using Rucio's "child rule" mechanism to link the disk and tape rule.  Alternatively if we had a way to query or remember the tape rules made for this dataset, we could check manually.  Either solution would be fine.

**Describe alternatives you've considered**
We could ignore the possibility of accidental deletion of tape rule source data and just hope that tape rules always get approved before the data disappears.  We could also lobby the Rucio development team to change the behavior so that unapproved tape rules still have some kind of data protection.  This has its own can of worms though.

**Additional context**
We also discussed the possibility of trying to catch rejected tape rules here, but I don't think we should get hung up on that.

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #10027 

#### Status
ready

#### Description
Almost all workflows, which are having some output, need to have tape rule created by MSOutput (account: `wmcore_output`). Exceptions are:
   * RelVals
   * Workflows configured in Unified to not save output on tape 

Up to now we were not accounting for tape rule status in the process of cleaning the Rucio rules for a given workflow. With the current change the status of all tape rules is required to be in a final ('OK') state, or no tape rules to have ever been created for the workflow, in order for the cleaning process to start.     

#### Is it backward compatible (if not, which system it affects?)
YES 

#### Related PRs
NO

#### External dependencies / deployment changes
NO

### CODE REVIEW

**Review by amaltaro — COMMENTED** [review]
Todor, this logic seems to be in the right direction.

In addition to the comments made along the code, I'd like to ask you to create that logging line when a workflow gets archived (as we have already discussed in the past). I know it's unrelated to this change, but it's one line and an important line for debugging ;)

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:183]: > I'm just trying to decrease the number of flags (checks)

I know, and me myself came with the same idea, but when I tried it, I figured out two things:
   * We take this flag from MSOutput as it is and it is set there by it's own logic, which is: 'Set the transfers to done once we have created the Rucio transfer rules, and do not care about the status of the transfers itself. Which is exactly the opposite of the logic we use here.
   * In case we decide to anyway reuse this flag here, we will end up with having a hidden difference in the logic reflected by the same flag in two different, but related services. And the flag  itself will constantly have different values in those two services - which will only cause confusion and will create situations for faulty debugging and conclusions.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:592]: Got it. 
Done.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:607]: Good point.

Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:607]: Done

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:183]: That's a valid point. Using the same key name/parameter with different meanings might be confusing indeed. On the other hand, these are different microservices and they have their own object data structure (MSOutputWflow vs MSRuleCleanerWflow), so one can also say their meaning is likely going to be different.

I have no strong preference though. Classical words: whatever is more robust, simple and maintainable ;-)

Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py:183]: Thanks Alan! in this case I would prefer to keep those flags separated. It also helps in the logging process. It gives clear reason what is the type of showstopper for the workflow to not being able to proceed: either a delay in the MSOutput to create the transfer document and rules or we are in the stage of waiting for a tape rule approval at a site which requires such.

**Comment by todor-ivanov** [comment]
Hi @amaltaro, I just  tested the change, by creating an artificial record in MongoDB for one of the workflows in my VM. I chose two 'TapeRuleId's  in two different states from a real workflow in prod, so that the checks should be triggered. And indeed they were, and also the cleaning of the workflow was not triggered exactly because on of the tape rules was not in a final state. 

So please go ahead and take another look at the latest version.    

p.s. A question came to my head wile working, but for that I will make a comment inline.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py]: @amaltaro with my latest commit I addressed the issu I was talking about here. Please take a look

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:613]: I would suggest to remove `or wrong Rucio server configured`. This should never be the case, unless it's in our private nodes and we are messing with things.

**Review by amaltaro — COMMENTED** [review]
Todor, code is looking good to me. However, I have made a couple of comments along the code where it might be worth it updating the code. Thanks

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:597]: Updating my own comment, it's actually not O(n^2). However, I still think it's a valid suggestion above.

Comment (amaltaro) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:597]: Thinking again, just disregard this comment, not worth it! Sorry for the noise.

**Review by amaltaro — COMMENTED** [review]
Todor, can you please squash those commits accordingly? Thanks

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:597]: Thanks @amaltaro 
>  to avoid O(n^2) complexity, how about changing the code to something like:

I can not see what you mean by this, because if you refer to the way how I build the list for the loop, I am quite sure it is `O(n)`. The list comprehension happens in place and is executed only once at the beginning of the `for` cycle, and since then it is kept in memory as a pure list for the consequent iterations. This is at least my understanding, and just to confirm it I did step through that cycle few times with pdb and it did build the list in advance. But I agree with you, for the sake of shortening the line and improving readability your suggestion has quite of a value. And also for not keeping in memory a list (even a short one) of something to which we already have a chain of references in memory. I am about to use it.

The other two comments you made here I really liked and I did manage to reduce the depth of `if` conditions here. Thanks  for that too.

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py]: Done

File: src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py]: Done

**Comment by todor-ivanov** [comment]
Thank you @amaltaro for your previous review and suggestions. I do hate nesting `if` conditions too deep. Now the code looks better to me also. Could you please take a look.

**Review by amaltaro — APPROVED** [review]
Code looks pretty good to me, Todor. Thanks!

**Pull request merged**
